File: prayer_app_expo/src/services/notificationService.js
import * as Notifications from 'expo-notifications';
import { Platform } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';

// Configure how notifications appear when app is in foreground
Notifications.setNotificationHandler({
    handleNotification: async () => ({
        shouldShowAlert: true,
        shouldPlaySound: true,
        shouldSetBadge: false,
    }),
});

const CHANNEL_ID = 'prayer_times';
const CACHE_KEY = 'salah_notif_cache';
const LAST_SCHEDULED_KEY = 'notif_last_scheduled';
let initialized = false;

// Debounce + single-flight state
let _debounceTimer = null;
let _inFlight = false;
let _pendingReason = null;

/** Initialize notification channel (Android) */
async function init() {
    if (initialized) return;

    if (Platform.OS === 'android') {
        await Notifications.setNotificationChannelAsync(CHANNEL_ID, {
            name: 'Prayer Times',
            description: 'Notifications for prayer times',
            importance: Notifications.AndroidImportance.HIGH,
            sound: 'default',
        });
    }

    initialized = true;
    console.log('[NOTIF] initialized');
}

/** Request permission. Returns true if granted. */
async function requestPermission() {
    const { status: existingStatus } = await Notifications.getPermissionsAsync();
    if (existingStatus === 'granted') {
        console.log('[NOTIF] permission granted');
        return true;
    }

    const { status } = await Notifications.requestPermissionsAsync({
        ios: { allowAlert: true, allowSound: true, allowBadge: true },
    });
    const granted = status === 'granted';
    console.log(`[NOTIF] permission ${granted ? 'granted' : 'denied'}`);
    return granted;
}

/** Send a test notification immediately */
async function sendTestNow() {
    await init();
    await Notifications.scheduleNotificationAsync({
        content: {
            title: 'Test Notification',
            body: 'Notifications are working ✅',
            sound: 'default',
            ...(Platform.OS === 'android' ? { channelId: CHANNEL_ID } : {}),
        },
        trigger: null,
    });
    console.log('[NOTIF] sent test now');
}

/** Schedule a test notification in 10 seconds */
async function scheduleTestIn10s() {
    await init();
    await Notifications.scheduleNotificationAsync({
        content: {
            title: 'Test Notification',
            body: 'Notifications are working ✅',
            sound: 'default',
            ...(Platform.OS === 'android' ? { channelId: CHANNEL_ID } : {}),
        },
        trigger: {
            type: Notifications.SchedulableTriggerInputTypes.TIME_INTERVAL,
            seconds: 10,
        },
    });
    console.log('[NOTIF] scheduled test in 10s');
}

/** Cache week timings for notification scheduling */
async function cacheTimingsForNotifications(serialized) {
    try {
        await AsyncStorage.setItem(CACHE_KEY, JSON.stringify(serialized));
        console.log(`[NOTIF] cached timings (${Object.keys(serialized).length} days)`);
    } catch (_) { /* skip */ }
}

/** Load cached timings */
async function _loadCachedTimings() {
    try {
        const raw = await AsyncStorage.getItem(CACHE_KEY);
        return raw ? JSON.parse(raw) : null;
    } catch (_) {
        return null;
    }
}

/**
 * Central entry point for (re)scheduling prayer notifications.
 * Debounced (300ms) + single-flight guard so multiple rapid calls collapse into one.
 * @param {string} reason - why this reschedule was triggered (for logging)
 */
function scheduleFromCache(reason = 'unknown') {
    // If already debouncing, update the reason and reset timer
    if (_debounceTimer) {
        clearTimeout(_debounceTimer);
        console.log(`[NOTIF] reschedule debounced (reason=${reason})`);
    }
    _pendingReason = reason;

    _debounceTimer = setTimeout(async () => {
        _debounceTimer = null;
        await _executeSchedule(_pendingReason);
    }, 300);
}

/** Actually execute the scheduling (called after debounce) */
async function _executeSchedule(reason) {
    // Single-flight: if already running, skip
    if (_inFlight) {
        console.log(`[NOTIF] schedule already in-flight, skipping (reason=${reason})`);
        return;
    }
    _inFlight = true;

    try {
        await init();

        // Load notification settings
        const enabledRaw = await AsyncStorage.getItem('notif_enabled');
        const enabled = enabledRaw === 'true';
        if (!enabled) {
            console.log(`[NOTIF] master toggle OFF, skipping schedule (reason=${reason})`);
            return;
        }

        const leadRaw = await AsyncStorage.getItem('notif_lead_minutes');
        const leadMinutes = leadRaw ? parseInt(leadRaw, 10) : 0;

        const enabledPrayers = {};
        for (const name of ['Fajr', 'Dhuhr', 'Asr', 'Maghrib', 'Isha']) {
            const val = await AsyncStorage.getItem(`notif_prayer_${name.toLowerCase()}`);
            enabledPrayers[name] = val === null ? true : val === 'true';
        }

        const cached = await _loadCachedTimings();
        if (!cached) {
            console.log(`[NOTIF] no cached timings, skipping (reason=${reason})`);
            return;
        }

        console.log(`[NOTIF] reschedule reason=${reason}`);
        await _scheduleFromParsed({ parsedTimings: cached, enabledPrayers, leadMinutes });
    } catch (e) {
        console.log('[NOTIF] scheduleFromCache error:', e);
    } finally {
        _inFlight = false;
    }
}

/** Internal: schedule from parsed timing data.
 *  Policy: schedule ALL enabled prayers within a rolling 48h window. */
async function _scheduleFromParsed({ parsedTimings, enabledPrayers, leadMinutes }) {
    await init();

    // Cancel only prayer notifications (identifiers starting with 'prayer_')
    const allScheduled = await Notifications.getAllScheduledNotificationsAsync();
    const prayerIds = allScheduled.filter(n => n.identifier?.startsWith('prayer_')).map(n => n.identifier);
    for (const id of prayerIds) {
        await Notifications.cancelScheduledNotificationAsync(id);
    }
    console.log(`[NOTIF] cancelled ${prayerIds.length} prayer notifications`);

    const now = new Date();
    const windowStartMs = now.getTime() + 5000; // now + 5s
    const windowEndMs = now.getTime() + 48 * 60 * 60 * 1000; // now + 48h
    const windowStart = new Date(windowStartMs);
    const windowEnd = new Date(windowEndMs);

    const offsetH = -(now.getTimezoneOffset() / 60);
    const offsetSign = offsetH >= 0 ? '+' : '';
    console.log(`[NOTIF] windowStart=${_fmtDt(windowStart)} windowEnd=${_fmtDt(windowEnd)} (${offsetSign}${offsetH}00)`);

    const scheduledMap = {};
    let scheduledCount = 0;

    // Collect all candidate triggers
    const candidates = [];

    for (const [dateKey, dayData] of Object.entries(parsedTimings)) {
        if (!dayData) continue;
        const prayers = dayData.prayers || [];

        for (const prayer of prayers) {
            const name = prayer.name;
            const time24 = prayer.time24;
            if (!enabledPrayers[name]) continue;

            // Parse time24 (HH:mm)
            const timeParts = time24?.split(':');
            if (!timeParts || timeParts.length < 2) continue;
            const hour = parseInt(timeParts[0], 10);
            const min = parseInt(timeParts[1], 10);

            // Parse dateKey (YYYY-MM-DD)
            const dateParts = dateKey.split('-');
            if (dateParts.length < 3) continue;
            const year = parseInt(dateParts[0], 10);
            const month = parseInt(dateParts[1], 10) - 1;
            const day = parseInt(dateParts[2], 10);

            const triggerTime = new Date(year, month, day, hour, min, 0);
            triggerTime.setMinutes(triggerTime.getMinutes() - leadMinutes);

            // Filter: must be within [windowStart, windowEnd]
            if (triggerTime.getTime() < windowStartMs) continue;
            if (triggerTime.getTime() > windowEndMs) continue;

            candidates.push({ name, time24, triggerTime, dateKey });
        }
    }

    // Sort by trigger time for deterministic ordering
    candidates.sort((a, b) => a.triggerTime.getTime() - b.triggerTime.getTime());

    // Schedule each candidate
    for (const c of candidates) {
        const { name, time24, triggerTime, dateKey } = c;

        const time12 = formatTo12Hour(time24);
        let body;
        if (leadMinutes > 0) {
            body = `${name} in ${leadMinutes} min • Adhan at ${time12}`;
        } else {
            body = `Adhan at ${time12}`;
        }

        const identifier = `prayer_${dateKey}_${name}`;
        const triggerStr = _fmtDt(triggerTime);

        await Notifications.scheduleNotificationAsync({
            identifier,
            content: {
                title: `${name} Prayer`,
                body,
                sound: 'default',
                ...(Platform.OS === 'android' ? { channelId: CHANNEL_ID } : {}),
            },
            trigger: {
                type: Notifications.SchedulableTriggerInputTypes.DATE,
                date: triggerTime,
            },
        });

        console.log(`[NOTIF] id=${identifier} prayer=${name} trigger=${triggerStr} (${offsetSign}${offsetH}00)`);

        scheduledMap[identifier] = {
            prayer: name,
            trigger: triggerStr,
            body,
            dateKey,
        };
        scheduledCount++;
    }

    // Persist for debug display
    try {
        await AsyncStorage.setItem(LAST_SCHEDULED_KEY, JSON.stringify(scheduledMap));
    } catch (_) { /* skip */ }

    console.log(`[NOTIF] scheduledCount=${scheduledCount}`);
}

/** Format Date as YYYY-MM-DD HH:mm */
function _fmtDt(dt) {
    return `${dt.getFullYear()}-${String(dt.getMonth()+1).padStart(2,'0')}-${String(dt.getDate()).padStart(2,'0')} ${String(dt.getHours()).padStart(2,'0')}:${String(dt.getMinutes()).padStart(2,'0')}`;
}

/** Get pending prayer notifications for debug.
 *  Falls back to stored schedule map if Expo Go hides trigger details. */
async function getPendingPrayerNotifications() {
    await init();

    // Load stored schedule map as fallback
    let storedMap = {};
    try {
        const raw = await AsyncStorage.getItem(LAST_SCHEDULED_KEY);
        if (raw) storedMap = JSON.parse(raw);
    } catch (_) { /* skip */ }

    const all = await Notifications.getAllScheduledNotificationsAsync();
    const pendingIds = new Set(
        all.filter(n => n.identifier?.startsWith('prayer_')).map(n => n.identifier)
    );

    // Build result from stored map, filtered to still-pending
    const result = [];
    for (const [id, info] of Object.entries(storedMap)) {
        if (!pendingIds.has(id)) continue;
        result.push({
            id,
            prayer: info.prayer || '',
            trigger: info.trigger || 'unknown',
            body: info.body || '',
        });
    }

    // Sort by trigger for consistent ordering
    result.sort((a, b) => a.trigger.localeCompare(b.trigger));

    console.log(`[NOTIF-DEBUG] pending prayer notifications: ${result.length}`);
    for (const r of result) {
        console.log(`[NOTIF-DEBUG]   id=${r.id} prayer=${r.prayer} trigger=${r.trigger}`);
    }
    return result;
}

/** Schedule a pipeline test: fires in 60 seconds using real pipeline */
async function schedulePipelineTestIn60s() {
    await init();

    try {
        await Notifications.cancelScheduledNotificationAsync('pipeline_test');
    } catch (_) { /* may not exist */ }

    const fireAt = new Date(Date.now() + 60 * 1000);

    await Notifications.scheduleNotificationAsync({
        identifier: 'pipeline_test',
        content: {
            title: 'Pipeline Test',
            body: 'Prayer notification pipeline works \u2705 (scheduled 60s ago)',
            sound: 'default',
            ...(Platform.OS === 'android' ? { channelId: CHANNEL_ID } : {}),
        },
        trigger: {
            type: Notifications.SchedulableTriggerInputTypes.DATE,
            date: fireAt,
        },
    });

    const offsetH = -(fireAt.getTimezoneOffset() / 60);
    const offsetSign = offsetH >= 0 ? '+' : '';
    const dateStr = `${fireAt.getFullYear()}-${String(fireAt.getMonth()+1).padStart(2,'0')}-${String(fireAt.getDate()).padStart(2,'0')} ${String(fireAt.getHours()).padStart(2,'0')}:${String(fireAt.getMinutes()).padStart(2,'0')}:${String(fireAt.getSeconds()).padStart(2,'0')}`;
    console.log(`[NOTIF] pipeline test scheduled at ${dateStr} (${offsetSign}${offsetH}00) id=pipeline_test`);
}

/** Cancel all scheduled notifications */
async function cancelAll() {
    await Notifications.cancelAllScheduledNotificationsAsync();
    console.log('[NOTIF] cancelled all');
}

/** Cancel only prayer notifications */
async function cancelPrayerNotifications() {
    const all = await Notifications.getAllScheduledNotificationsAsync();
    const prayerIds = all.filter(n => n.identifier?.startsWith('prayer_')).map(n => n.identifier);
    for (const id of prayerIds) {
        await Notifications.cancelScheduledNotificationAsync(id);
    }
    console.log(`[NOTIF] cancelled prayer notifications count=${prayerIds.length}`);
}

/** Helper: HH:mm -> h:mm AM/PM */
function formatTo12Hour(time24) {
    if (!time24) return '';
    const [h, m] = time24.split(':').map(Number);
    const ampm = h >= 12 ? 'PM' : 'AM';
    const hour12 = h % 12 || 12;
    return `${hour12}:${String(m).padStart(2, '0')} ${ampm}`;
}

export default {
    init,
    requestPermission,
    sendTestNow,
    scheduleTestIn10s,
    cacheTimingsForNotifications,
    scheduleFromCache,
    getPendingPrayerNotifications,
    schedulePipelineTestIn60s,
    cancelAll,
    cancelPrayerNotifications,
};

File: prayer_app_flutter/lib/src/services/notification_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:timezone/data/latest_all.dart' as tz;
import 'package:timezone/timezone.dart' as tz;

class NotificationService {
  static final NotificationService _instance = NotificationService._();
  factory NotificationService() => _instance;
  NotificationService._();

  final FlutterLocalNotificationsPlugin _plugin =
      FlutterLocalNotificationsPlugin();

  bool _initialized = false;

  static const _channelId = 'prayer_times';
  static const _channelName = 'Prayer Times';
  static const _channelDesc = 'Notifications for prayer times';
  static const _cacheKey = 'salah_notif_cache';

  /// Initialize plugin + timezone + request permission
  Future<void> init() async {
    if (_initialized) return;

    tz.initializeTimeZones();

    // TIMEZONE STRATEGY:
    // We use the device's local timezone for scheduling notifications.
    // The prayer times in cache are computed by the API for the chosen location.
    // If device timezone != location timezone (e.g., device in EET but location
    // set to San Francisco PST), the scheduled notification times will be wrong.
    // This is expected in dev/testing. In production, users are at their location.
    final tzName = _resolveTimezone();
    tz.setLocalLocation(tz.getLocation(tzName));
    debugPrint('[NOTIF] timezone resolved to: $tzName (device offset: ${DateTime.now().timeZoneOffset})');

    const androidInit = AndroidInitializationSettings('@mipmap/ic_launcher');
    const iosInit = DarwinInitializationSettings(
      requestAlertPermission: true,
      requestSoundPermission: true,
      requestBadgePermission: true,
    );
    const settings = InitializationSettings(
      android: androidInit,
      iOS: iosInit,
    );

    await _plugin.initialize(settings: settings);

    // Create Android channel
    const androidChannel = AndroidNotificationChannel(
      _channelId,
      _channelName,
      description: _channelDesc,
      importance: Importance.high,
    );
    await _plugin
        .resolvePlatformSpecificImplementation<
            AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannel(androidChannel);

    _initialized = true;
    debugPrint('[NOTIF] initialized');
  }

  /// Request permission. Returns true if granted.
  Future<bool> requestPermission() async {
    if (Platform.isIOS) {
      final result = await _plugin
          .resolvePlatformSpecificImplementation<
              IOSFlutterLocalNotificationsPlugin>()
          ?.requestPermissions(alert: true, sound: true, badge: true);
      final granted = result ?? false;
      debugPrint('[NOTIF] permission ${granted ? "granted" : "denied"}');
      return granted;
    }
    if (Platform.isAndroid) {
      final androidPlugin = _plugin.resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin>();

      // Request exact alarm permission for Android 12+
      await androidPlugin?.requestExactAlarmsPermission();

      final granted = await androidPlugin?.requestNotificationsPermission();
      debugPrint(
          '[NOTIF] permission ${(granted ?? true) ? "granted" : "denied"}');
      return granted ?? true;
    }
    return true;
  }

  /// Send a test notification immediately
  Future<void> sendTestNow() async {
    await init();
    const details = NotificationDetails(
      android: AndroidNotificationDetails(
        _channelId,
        _channelName,
        channelDescription: _channelDesc,
        importance: Importance.high,
        priority: Priority.high,
      ),
      iOS: DarwinNotificationDetails(),
    );
    await _plugin.show(
      id: 0,
      title: 'Test Notification',
      body: 'Notifications are working ✅',
      notificationDetails: details,
    );
    debugPrint('[NOTIF] sent test now');
  }

  /// Schedule a test notification in 10 seconds
  Future<void> scheduleTestIn10s() async {
    await init();
    const details = NotificationDetails(
      android: AndroidNotificationDetails(
        _channelId,
        _channelName,
        channelDescription: _channelDesc,
        importance: Importance.high,
        priority: Priority.high,
      ),
      iOS: DarwinNotificationDetails(),
    );
    await _plugin.zonedSchedule(
      id: 1,
      title: 'Test Notification',
      body: 'Notifications are working ✅',
      scheduledDate: tz.TZDateTime.now(tz.local).add(const Duration(seconds: 10)),
      notificationDetails: details,
      androidScheduleMode: AndroidScheduleMode.exactAllowWhileIdle,
    );
    debugPrint('[NOTIF] scheduled test in 10s');
  }

  /// Cache week timings to persistent storage for use by settings reschedule
  Future<void> cacheTimingsForNotifications(Map<String, dynamic> serialized) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_cacheKey, jsonEncode(serialized));
    debugPrint('[NOTIF] cached timings for notifications (${serialized.length} days)');
  }

  /// Load cached timings
  Future<Map<String, dynamic>?> _loadCachedTimings() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_cacheKey);
    if (raw == null) return null;
    return jsonDecode(raw) as Map<String, dynamic>;
  }

  /// Schedule prayer notifications using cached timings + current settings
  Future<void> scheduleFromCache() async {
    await init();
    final prefs = await SharedPreferences.getInstance();

    // Load notification settings
    final enabled = prefs.getBool('notif_enabled') ?? false;
    if (!enabled) {
      debugPrint('[NOTIF] master toggle OFF, skipping schedule');
      return;
    }

    final leadMinutes = prefs.getInt('notif_lead_minutes') ?? 0;
    final enabledPrayers = <String, bool>{};
    for (final name in ['Fajr', 'Dhuhr', 'Asr', 'Maghrib', 'Isha']) {
      enabledPrayers[name] = prefs.getBool('notif_prayer_${name.toLowerCase()}') ?? true;
    }

    final cached = await _loadCachedTimings();
    if (cached == null) {
      debugPrint('[NOTIF] no cached timings, skipping schedule');
      return;
    }

    await _scheduleFromParsed(
      parsedTimings: cached,
      enabledPrayers: enabledPrayers,
      leadMinutes: leadMinutes,
    );
  }

  /// Schedule prayer notifications from parsed timing data.
  /// Policy: schedule ALL enabled prayers within a rolling 48h window.
  Future<void> _scheduleFromParsed({
    required Map<String, dynamic> parsedTimings,
    required Map<String, bool> enabledPrayers,
    required int leadMinutes,
  }) async {
    await init();

    // Cancel only prayer notifications (IDs 100+), keep test notifs (0, 1, 99)
    final pending = await _plugin.pendingNotificationRequests();
    final prayerIds = pending.where((p) => p.id >= 100).map((p) => p.id).toList();
    for (final id in prayerIds) {
      await _plugin.cancel(id: id);
    }
    debugPrint('[NOTIF] cancelled ${prayerIds.length} prayer notifications');

    final now = DateTime.now();
    final windowStart = now.add(const Duration(seconds: 5));
    final windowEnd = now.add(const Duration(hours: 48));

    final offsetH = now.timeZoneOffset.inHours;
    final offsetSign = offsetH >= 0 ? '+' : '';
    final windowStartStr = _fmtDt(windowStart);
    final windowEndStr = _fmtDt(windowEnd);
    debugPrint('[NOTIF] windowStart=$windowStartStr windowEnd=$windowEndStr ($offsetSign${offsetH}00)');

    int notifId = 100;
    int scheduledCount = 0;
    final scheduledList = <Map<String, String>>[];

    const details = NotificationDetails(
      android: AndroidNotificationDetails(
        _channelId,
        _channelName,
        channelDescription: _channelDesc,
        importance: Importance.high,
        priority: Priority.high,
      ),
      iOS: DarwinNotificationDetails(),
    );

    // Collect all candidate triggers, then sort by time
    final candidates = <Map<String, dynamic>>[];

    for (final entry in parsedTimings.entries) {
      final dateKey = entry.key;
      final dayData = entry.value;
      if (dayData == null) continue;

      final prayers = dayData['prayers'] as List<dynamic>?;
      if (prayers == null) continue;

      for (final prayer in prayers) {
        final name = prayer['name'] as String;
        final time24 = prayer['time24'] as String;

        if (enabledPrayers[name] != true) continue;

        // Parse time24 (HH:mm)
        final timeParts = time24.split(':');
        if (timeParts.length < 2) continue;
        final hour = int.tryParse(timeParts[0]) ?? 0;
        final min = int.tryParse(timeParts[1]) ?? 0;

        // Parse dateKey (YYYY-MM-DD)
        final dateParts = dateKey.split('-');
        if (dateParts.length < 3) continue;
        final year = int.tryParse(dateParts[0]) ?? now.year;
        final month = int.tryParse(dateParts[1]) ?? now.month;
        final day = int.tryParse(dateParts[2]) ?? now.day;

        var triggerTime = DateTime(year, month, day, hour, min);
        triggerTime = triggerTime.subtract(Duration(minutes: leadMinutes));

        // Filter: must be within [windowStart, windowEnd]
        if (triggerTime.isBefore(windowStart)) continue;
        if (triggerTime.isAfter(windowEnd)) continue;

        candidates.add({
          'name': name,
          'time24': time24,
          'triggerTime': triggerTime,
          'dateKey': dateKey,
        });
      }
    }

    // Sort by trigger time for deterministic ordering
    candidates.sort((a, b) =>
        (a['triggerTime'] as DateTime).compareTo(b['triggerTime'] as DateTime));

    // Schedule each candidate
    for (final c in candidates) {
      final name = c['name'] as String;
      final time24 = c['time24'] as String;
      final triggerTime = c['triggerTime'] as DateTime;
      final dateKey = c['dateKey'] as String;

      final tzTime = tz.TZDateTime.from(triggerTime, tz.local);

      final time12 = _formatTo12Hour(time24);
      String body;
      if (leadMinutes > 0) {
        body = '$name in $leadMinutes min • Adhan at $time12';
      } else {
        body = 'Adhan at $time12';
      }

      await _plugin.zonedSchedule(
        id: notifId,
        title: '$name Prayer',
        body: body,
        scheduledDate: tzTime,
        notificationDetails: details,
        androidScheduleMode: AndroidScheduleMode.exactAllowWhileIdle,
      );

      final triggerStr = _fmtDt(triggerTime);
      debugPrint('[NOTIF] id=$notifId prayer=$name trigger=$triggerStr ($offsetSign${offsetH}00)');

      scheduledList.add({
        'id': '$notifId',
        'prayer': name,
        'trigger': triggerStr,
        'body': body,
        'dateKey': dateKey,
      });

      notifId++;
      scheduledCount++;
    }

    // Persist schedule list for debug display
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('notif_last_scheduled', jsonEncode(scheduledList));

    debugPrint('[NOTIF] scheduledCount=$scheduledCount');
  }

  /// Get pending prayer notifications for debug inspection.
  /// Uses stored schedule list for trigger times (OS doesn't expose them).
  Future<List<Map<String, String>>> getPendingPrayerNotifications() async {
    await init();

    // Load stored schedule list
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString('notif_last_scheduled');
    List<Map<String, String>> stored = [];
    if (raw != null) {
      final decoded = jsonDecode(raw) as List<dynamic>;
      stored = decoded.map((e) => Map<String, String>.from(e as Map)).toList();
    }

    final pending = await _plugin.pendingNotificationRequests();
    final pendingIds = pending.where((p) => p.id >= 100).map((p) => '${p.id}').toSet();

    // Filter stored list to only include items still pending
    final result = stored.where((s) => pendingIds.contains(s['id'])).toList();

    debugPrint('[NOTIF-DEBUG] pending prayer notifications: ${result.length}');
    for (final r in result) {
      debugPrint('[NOTIF-DEBUG]   id=${r['id']} prayer=${r['prayer']} trigger=${r['trigger']}');
    }
    return result;
  }

  /// Schedule a pipeline test: uses the real prayer notification pipeline
  /// but fires in 60 seconds. This verifies the full scheduling path works.
  Future<void> schedulePipelineTestIn60s() async {
    await init();

    // Cancel any previous pipeline test (ID 99)
    await _plugin.cancel(id: 99);

    final now = DateTime.now();
    final fireAt = now.add(const Duration(seconds: 60));
    final tzTime = tz.TZDateTime.from(fireAt, tz.local);

    const details = NotificationDetails(
      android: AndroidNotificationDetails(
        _channelId,
        _channelName,
        channelDescription: _channelDesc,
        importance: Importance.high,
        priority: Priority.high,
      ),
      iOS: DarwinNotificationDetails(),
    );

    await _plugin.zonedSchedule(
      id: 99,
      title: 'Pipeline Test',
      body: 'Prayer notification pipeline works ✅ (scheduled 60s ago)',
      scheduledDate: tzTime,
      notificationDetails: details,
      androidScheduleMode: AndroidScheduleMode.exactAllowWhileIdle,
    );

    final offsetH = fireAt.timeZoneOffset.inHours;
    final offsetSign = offsetH >= 0 ? '+' : '';
    debugPrint('[NOTIF] pipeline test scheduled at ${fireAt.year}-${fireAt.month.toString().padLeft(2, '0')}-${fireAt.day.toString().padLeft(2, '0')} ${fireAt.hour.toString().padLeft(2, '0')}:${fireAt.minute.toString().padLeft(2, '0')}:${fireAt.second.toString().padLeft(2, '0')} ($offsetSign${offsetH}00) id=99');
  }

  /// Cancel all notifications
  Future<void> cancelAll() async {
    await _plugin.cancelAll();
    debugPrint('[NOTIF] cancelled all');
  }

  /// Cancel only prayer notifications (keep test ones)
  Future<void> cancelPrayerNotifications() async {
    final pending = await _plugin.pendingNotificationRequests();
    int count = 0;
    for (final p in pending) {
      if (p.id >= 100) {
        await _plugin.cancel(id: p.id);
        count++;
      }
    }
    debugPrint('[NOTIF] cancelled prayer notifications count=$count');
  }

  String _resolveTimezone() {
    // Map common abbreviations to tz database names
    final tzName = DateTime.now().timeZoneName;
    // Try direct lookup first
    try {
      tz.getLocation(tzName);
      return tzName;
    } catch (_) {}

    // Common mappings
    final offset = DateTime.now().timeZoneOffset;
    final hours = offset.inHours;

    // Map by offset as fallback
    final offsetMap = <int, String>{
      -12: 'Etc/GMT+12',
      -11: 'Pacific/Pago_Pago',
      -10: 'Pacific/Honolulu',
      -9: 'America/Anchorage',
      -8: 'America/Los_Angeles',
      -7: 'America/Denver',
      -6: 'America/Chicago',
      -5: 'America/New_York',
      -4: 'America/Santiago',
      -3: 'America/Sao_Paulo',
      -2: 'Atlantic/South_Georgia',
      -1: 'Atlantic/Azores',
      0: 'Europe/London',
      1: 'Europe/Paris',
      2: 'Europe/Bucharest',
      3: 'Europe/Moscow',
      4: 'Asia/Dubai',
      5: 'Asia/Karachi',
      6: 'Asia/Dhaka',
      7: 'Asia/Bangkok',
      8: 'Asia/Shanghai',
      9: 'Asia/Tokyo',
      10: 'Australia/Sydney',
      11: 'Pacific/Noumea',
      12: 'Pacific/Auckland',
    };

    return offsetMap[hours] ?? 'UTC';
  }

  String _formatTo12Hour(String time24) {
    final parts = time24.split(':');
    if (parts.length < 2) return time24;
    final h = int.tryParse(parts[0]) ?? 0;
    final m = int.tryParse(parts[1]) ?? 0;
    final ampm = h >= 12 ? 'PM' : 'AM';
    final hour12 = h % 12 == 0 ? 12 : h % 12;
    return '$hour12:${m.toString().padLeft(2, '0')} $ampm';
  }

  /// Format DateTime as YYYY-MM-DD HH:mm
  String _fmtDt(DateTime dt) {
    return '${dt.year}-${dt.month.toString().padLeft(2, '0')}-${dt.day.toString().padLeft(2, '0')} '
        '${dt.hour.toString().padLeft(2, '0')}:${dt.minute.toString().padLeft(2, '0')}';
  }
}
